pkg/collector/corechecks/servicediscovery/impl_linux.go

```diff
@@ -85,120 +81,53 @@ func (li *linuxImpl) DiscoverServices() (*discoveredServices, error) {
 		}
 	}
 
-	// The endpoint could be refactored in the future to return a map to avoid this.
-	serviceMap := make(map[int]*model.Service, len(response.Services))
-	for _, service := range response.Services {
-		serviceMap[service.PID] = &service
-	}
-
 	events := serviceEvents{}
-	now := li.time.Now()
-
-	li.handlePotentialServices(&events, now, serviceMap)
 
-	// check open ports - these will be potential new services if they are still alive in the next iteration.
-	for _, service := range response.Services {
+	for _, service := range response.StartedServices {
 		pid := service.PID
 		if li.ignoreProcs[pid] {
 			continue
 		}
-		if _, ok := li.aliveServices[pid]; !ok {
-			log.Debugf("[pid: %d] found new process with open ports", pid)
-
-			svc := li.getServiceInfo(service)
-			if li.ignoreCfg[svc.meta.Name] {
-				log.Debugf("[pid: %d] process ignored from config: %s", pid, svc.meta.Name)
-				li.ignoreProcs[pid] = true
-				continue
-			}
 
-			log.Debugf("[pid: %d] adding process to potential: %s", pid, svc.meta.Name)
-			li.potentialServices[pid] = &svc
+		log.Debugf("[pid: %d] found new process with open ports", pid)
+		if li.ignoreCfg[service.Name] {
+			log.Debugf("[pid: %d] process ignored from config: %s", pid, service.Name)
+			li.ignoreProcs[pid] = true
+			continue
 		}
+
+		li.aliveServices[pid] = &service
+		events.start = append(events.start, service)
 	}
 
-	// check if services previously marked as alive still are.
-	for pid, svc := range li.aliveServices {
-		if service, ok := serviceMap[pid]; !ok {
-			delete(li.aliveServices, pid)
-			events.stop = append(events.stop, *svc)
-		} else if now.Sub(svc.LastHeartbeat).Truncate(time.Minute) >= heartbeatTime {
-			svc.LastHeartbeat = now
-			svc.service.RSS = service.RSS
-			svc.service.CPUCores = service.CPUCores
-			svc.service.ContainerID = service.ContainerID
-			svc.service.GeneratedName = service.GeneratedName
-			svc.service.ContainerServiceName = service.ContainerServiceName
-			svc.service.ContainerServiceNameSource = service.ContainerServiceNameSource
-			svc.service.Name = service.Name
-			svc.meta.Name = service.Name
-			events.heartbeat = append(events.heartbeat, *svc)
+	for _, service := range response.StoppedServices {
+		if li.ignoreCfg[service.Name] {
```

Comment:
Looks like there is a leak here. If a service is ignored its pid will be added to `ignoreProcs` when it starts. But when it stops we won't get to the code below to delete the entry from `ignoreProcs` since we `continue` here.

Also, below, we delete the proc from the `ignoreProcs` but still send the stop event.

I think it's best to just get rid of the whole ignore procs thing from the check since there is another one in system-probe; I'll push that as a separate PR.